File: cc/google/fhir/r4/json_format.cc
/*
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "google/fhir/r4/json_format.h"

#include <string>

#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/strings/string_view.h"
#include "google/fhir/error_reporter.h"
#include "google/fhir/json/fhir_json.h"
#include "google/fhir/json_format.h"
#include "google/fhir/json_format_results.h"
#include "google/fhir/r4/primitive_handler.h"

namespace google {
namespace fhir {
namespace r4 {

namespace {

const Parser* GetParser() {
  static Parser* parser = new Parser(R4PrimitiveHandler::GetInstance());
  return parser;
}

const Printer* GetPrinter() {
  static Printer* printer = new Printer(R4PrimitiveHandler::GetInstance());
  return printer;
}

}  // namespace

absl::Status MergeJsonFhirStringIntoProto(const absl::string_view raw_json,
                                          google::protobuf::Message* target,
                                          absl::TimeZone default_timezone,
                                          const bool validate,
                                          ErrorHandler& error_handler) {
  FHIR_ASSIGN_OR_RETURN(
      ParseResult result,
      GetParser()->MergeJsonFhirStringIntoProto(
          raw_json, target, default_timezone, validate, error_handler));

  if (result == ParseResult::kFailed) {
    return absl::InvalidArgumentError(
        "Merge failure when parsing JSON.  See ErrorHandler for more info.");
  }
  return absl::OkStatus();
}

absl::Status MergeJsonFhirObjectIntoProto(
    const google::fhir::internal::FhirJson& json_object,
    google::protobuf::Message* target, absl::TimeZone default_timezone,
    const bool validate, ErrorHandler& error_handler) {
  FHIR_ASSIGN_OR_RETURN(
      ParseResult result,
      GetParser()->MergeJsonFhirObjectIntoProto(
          json_object, target, default_timezone, validate, error_handler));

  if (result == ParseResult::kFailed) {
    return absl::InvalidArgumentError(
        "Merge failure when parsing JSON.  See ErrorHandler for more info.");
  }
  return absl::OkStatus();
}

absl::StatusOr<std::string> PrintFhirPrimitive(
    const ::google::protobuf::Message& message) {
  return GetPrinter()->PrintFhirPrimitive(message);
}

absl::StatusOr<std::string> PrintFhirToJsonString(
    const google::protobuf::Message& fhir_proto) {
  return GetPrinter()->PrintFhirToJsonString(fhir_proto);
}

absl::StatusOr<std::string> PrettyPrintFhirToJsonString(
    const google::protobuf::Message& fhir_proto) {
  return GetPrinter()->PrettyPrintFhirToJsonString(fhir_proto);
}

absl::StatusOr<std::string> PrintFhirToJsonStringForAnalytics(
    const google::protobuf::Message& fhir_proto) {
  return GetPrinter()->PrintFhirToJsonStringForAnalytics(fhir_proto);
}

absl::StatusOr<std::string> PrettyPrintFhirToJsonStringForAnalytics(
    const google::protobuf::Message& fhir_proto) {
  return GetPrinter()->PrettyPrintFhirToJsonStringForAnalytics(fhir_proto);
}

}  // namespace r4
}  // namespace fhir
}  // namespace google

File: go/jsonformat/internal/jsonpbhelper/json_format.go
// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package jsonpbhelper

import (
	"strings"
	"sync"
	"unicode"

	"google.golang.org/protobuf/reflect/protoreflect"
)

const (
	// ContainedField is the JSON field name of inline resources.
	ContainedField = "contained"
	// ResourceTypeField constant.
	ResourceTypeField = "resourceType"
	// OneofName field constant.
	OneofName = "oneof_resource"
	// Extension field constant.
	Extension = "extension"
)

// IsJSON defines JSON related interface.
type IsJSON interface {
	// IsJSON method.
	IsJSON()
}

// JSONObject defines JSON map.
type JSONObject map[string]IsJSON

// IsJSON implementation of JSON object.
func (JSONObject) IsJSON() {}

// JSONArray defines JSON array.
type JSONArray []IsJSON

// IsJSON implementation of JSON array.
func (JSONArray) IsJSON() {}

// JSONString defines JSON string.
type JSONString string

// IsJSON implementation of JSON string.
func (JSONString) IsJSON() {}

// JSONRawValue defines JSON bytes.
type JSONRawValue []byte

// IsJSON implementation of JSON bytes.
func (JSONRawValue) IsJSON() {}

// MarshalJSON marshals JSON to bytes.
func (v JSONRawValue) MarshalJSON() ([]byte, error) {
	return v, nil
}

var (
	nameMapCache sync.Map // type map[string]*map[string]string
)

// SnakeToCamel converts a snake_case string into a CamelCase string.
func SnakeToCamel(snakeCase string) string {
	camelCase := make([]rune, 0, len(snakeCase))
	upper := true
	for _, c := range snakeCase {
		if c == '_' {
			upper = true
		} else if upper {
			upper = false
			camelCase = append(camelCase, unicode.ToUpper(c))
		} else {
			camelCase = append(camelCase, unicode.ToLower(c))
		}
	}
	return string(camelCase)
}

// CamelToSnake converts a CamelCase string into snake_case by putting one '_' character before each uppercase char and lowercasing all characters.
func CamelToSnake(camelCase string) string {
	snakeCase := make([]rune, 0, len(camelCase))
	for _, c := range camelCase {
		if !unicode.IsLower(c) {
			if len(snakeCase) > 0 {
				snakeCase = append(snakeCase, '_')
			}
		}
		snakeCase = append(snakeCase, unicode.ToLower(c))
	}
	return string(snakeCase)
}

func buildNameMap(d protoreflect.MessageDescriptor) *map[string]string {
	m := &map[string]string{}
	fields := d.Fields()
	for i := 0; i < fields.Len(); i++ {
		f := fields.Get(i)
		if f.HasJSONName() {
			(*m)[CamelToSnake(f.JSONName())] = string(f.Name())
		}
	}
	return m
}

func getNameMap(d protoreflect.MessageDescriptor) *map[string]string {
	n := d.FullName()
	if v, ok := nameMapCache.Load(n); ok {
		return v.(*map[string]string)
	}
	v, _ := nameMapCache.LoadOrStore(n, buildNameMap(d))
	return v.(*map[string]string)
}

// GetField returns the field descriptor.
func GetField(d protoreflect.MessageDescriptor, f string) protoreflect.FieldDescriptor {
	s := CamelToSnake(f)

	if fd := d.Fields().ByName(protoreflect.Name(s)); fd != nil {
		return fd
	}
	// Try again, as there may be a field whose json_name matches fd.
	return d.Fields().ByName(protoreflect.Name((*getNameMap(d))[s]))
}

// SnakeToLowerCamel converts the snake case to lower camel case.
func SnakeToLowerCamel(s string) string {
	if len(s) == 0 {
		return s
	}
	camel := SnakeToCamel(s)
	return strings.ToLower(camel[:1]) + camel[1:]
}
